### PR TITLE
fix(iam): importedRoleStackSafeDefaultPolicyName feature flag results in excessively long IAM policy names

### LINKED ISSUE #27409: importedRoleStackSafeDefaultPolicyName feature flag results in excessively long IAM policy names
### Describe the bug

When using the @aws-cdk/aws-iam:importedRoleStackSafeDefaultPolicyName feature flag with the CDK pipeline, the CDK generates exceedingly long IAM policy names that exceed AWS's character limits.

### Expected Behavior

The CDK should generate names that respect AWS's character limits regardless of feature flag settings.

### Current Behavior

When this feature flag is enabled, I get an error indicating that the policy name has exceeded the maximum allowed length (128 characters). The error message is as follows:

CREATE_FAILED Properties validation failed for resource Route53CrossAccountZoneDelegationRecordcrossaccountzonedelegationhandlerrolePolicyCadPlatformtoolboxDevToolchainCdkPipelineDevCadPlatformDevRoute53CrossAccountZoneDelegationRecordcrossaccountzonedelegationhandlerrole195BFE8B8FD65C38 with message: #/PolicyName: expected maxLength: 128, actual: 148

### Reproduction Steps

Enable the @aws-cdk/aws-iam:importedRoleStackSafeDefaultPolicyName feature flag with the CDK pipeline,
Define a stack that imports an IAM role and attach policies to it.
Deploy the stack.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### CDK CLI Version

2.99.1 (build b2a895e)

### Framework Version

_No response_

### Node.js Version

v18.18.0

### OS

Ubuntu 23.04

### Language

Python

### Language Version

Python 3.11.5

### Other information

_No response_

### PR DESCRIPTION
**Pull request by yamoyamoto**
When the importedRoleStackSafeDefaultPolicyName feature flag is enabled, the method to calculate the IAM Policy Name within `aws_iam.ImportedRole.addToPrincipalPolicy()` changes. Specifically, if the generated IAM Policy Name exceeds the maximum allowed length of 128 characters, it will be truncated using `Names.uniqueResourceName()`.

Previously, the `Names.UniqueId()` method was used to generate the Policy Name. This method does not allow you to set a maximum length, so if the name exceeded the limit, it would be overwritten using `Names.uniqueResourceName()`—a function that allows for length specification.

I considered replacing `Names.UniqueId()` entirely with `Names.uniqueResourceName()`. However, this is on hold due to concerns that existing Policy Names could be affected. If a complete replacement poses no issues, your guidance is appreciated, as I'm not fully versed in the logic behind these methods.

Closes #27409 , #24441 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by kaizencc — DISMISSED** [review]
Looks good @yamoyamoto. Sorry it took us so long to get to this PR

File: packages/aws-cdk-lib/aws-iam/test/role.from-role-arn.test.ts [hunk 1]
```diff
@@ -78,6 +78,21 @@ describe('IAM Role.fromRoleArn', () => {
           expect(stack2PolicyNameCapture.asString()).toMatch(/PolicyRoleStack2ImportedRole.*/);
         });
 
+        test('Policy name is truncated to a maximum length of 128 characters when the original name exceeds this limit', () =>{
```
[the remark this answers is not in the document]
Comment (yamoyamoto) [reply packages/aws-cdk-lib/aws-iam/test/role.from-role-arn.test.ts]: Sorry, I updated this.

https://github.com/aws/aws-cdk/pull/27548/commits/0713874b4cb573994e097aa8cc85ec7bae1e0d76

File: packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts [hunk 2]
```diff
@@ -45,9 +46,13 @@ export class ImportedRole extends Resource implements IRole, IComparablePrincipa
...
       const useUniqueName = FeatureFlags.of(this).isEnabled(IAM_IMPORTED_ROLE_STACK_SAFE_DEFAULT_POLICY_NAME);
-      const defaultDefaultPolicyName = useUniqueName
-        ? `Policy${Names.uniqueId(this)}`
-        : 'Policy';
+      const prefix = 'Policy';
+      let defaultDefaultPolicyName = useUniqueName
+        ? `${prefix}${Names.uniqueId(this)}`
+        : prefix;
+      if (defaultDefaultPolicyName.length > MAX_POLICY_NAME_LEN) {
+        defaultDefaultPolicyName = `Policy${Names.uniqueResourceName(this, { maxLength: MAX_POLICY_NAME_LEN - prefix.length })}`;
```
[the remark this answers is not in the document]
Comment (yamoyamoto) [reply packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts]: Indeed, it does become a bit difficult to read. However, we need prefix.length to subtract the length of 'Policy' from the policy name limit when using Names.uniqueResourceName(), so it seems reasonable to keep it as a variable.

>  but if we're doing that then we should use it on line 54 also.

My apologies, it was a simple oversight. I have now used the prefix variable.

https://github.com/aws/aws-cdk/pull/27548/commits/c009d92df788840fbf1d8af76b8a114192b10e3f

File: packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts [hunk 3]
```diff
@@ -45,9 +46,13 @@ export class ImportedRole extends Resource implements IRole, IComparablePrincipa
   public addToPrincipalPolicy(statement: PolicyStatement): AddToPrincipalPolicyResult {
     if (!this.defaultPolicy) {
       const useUniqueName = FeatureFlags.of(this).isEnabled(IAM_IMPORTED_ROLE_STACK_SAFE_DEFAULT_POLICY_NAME);
-      const defaultDefaultPolicyName = useUniqueName
-        ? `Policy${Names.uniqueId(this)}`
-        : 'Policy';
+      const prefix = 'Policy';
```
[the remark this answers is not in the document]
Comment (yamoyamoto) [reply packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts:51]: I added comment. I would appreciate it if you could confirm.

https://github.com/aws/aws-cdk/pull/27548/commits/4850d8285edbd522a44f442abd9091af720080bf

**Review by kaizencc — APPROVED** [review]
Thanks @yamoyamoto!

**Review by kaizencc — APPROVED** [review]

**Pull request merged**